### PR TITLE
Implement undo

### PR DESCRIPTION
**Pull request by r4ulill0**
The undo functionality does not really resolve recurring todos because that functionality is not implemented yet so there is a test that will fail until that is implemented (I'm planing to solve that myself, but one thing at a time).

@WhyNotHugo tell me if you see something off and sorry for taking that long, recurring todos distracted me a bit :laughing:

### CODE REVIEW

File: todoman/model.py [hunk 1]
```diff
@@ -252,6 +252,27 @@ def complete(self) -> None:
         self.percent_complete = 100
         self.status = "COMPLETED"
 
+    def undo(self) -> Todo | None:
+        """
+        Immediately restores this todo. Marks it as needs action, resets the
+        percentage to 0 and deletes the completed_at datetime.
+
```
Comment (WhyNotHugo) [inline todoman/model.py:259]:
```suggestion
        """Immediately restores this todo.

        Marks it as needs action, resets the percentage to 0 and deletes the
        completed_at datetime.
```

The first line is show in short help output.

**Review by WhyNotHugo — APPROVED** [review]

**Comment by r4ulill0** [comment]
> I don't fully understand why one tests is failing: https://builds.sr.ht/~whynothugo/job/981248#task-test-262

Recurring test doesn't work because there are some recursion fields that are not persisted yet. We have to deal some issues  #485 in order for it to work. I'm planning on taking all of them but one step at a time.

I tried to implement persistence for RELATED-TO but it was a pretty big change and it has its own issue #288.
You can open a new issue to fix this test and add it to #485 if you want. Also we can delete the test until its relevant. I'm fine with either option.

**Comment by r4ulill0** [comment]
I have marked the test as the RELATED-TO property needs to be persisted before it can work.

Now arch build passes (Alpine build just exploded and I don't understand why).

**Comment by dikey0ficial** [comment]
Is there any progress?..

**Comment by r4ulill0** [comment]
> Is there any progress?..

It should be ready to merge. I don't know if @WhyNotHugo has any more things to check.

File: todoman/cli.py [hunk 2]
```diff
@@ -469,6 +469,30 @@ def done(ctx, todos):
...
+@cli.command()
+@pass_ctx
+@click.argument(
+    "todos",
+    nargs=-1,
+    required=True,
+    type=click.IntRange(0),
+    callback=_validate_todos,
+)
+# @catch_errors
```
Comment (WhyNotHugo) [inline todoman/cli.py:481]: Is this bit problematic?

File: todoman/model.py [hunk 3]
```diff
@@ -250,6 +250,27 @@ def complete(self) -> None:
...
 
+    def undo(self) -> Todo | None:
+        """Immediately restores this todo.
+
+        Marks it as needs action, resets the percentage to 0 and deletes the
+        completed_at datetime.
+        If this todo belongs to a series, the one created at completion will
+        be deleted.
+
+        Returns the todo that will be deleted.
```
Comment (WhyNotHugo) [inline todoman/model.py:261]:
```suggestion
        Returns the todo that should be deleted.
```

File: todoman/model.py [hunk 4]
```diff
@@ -250,6 +250,27 @@ def complete(self) -> None:
...
+        """
+        original = None
+        if self.is_recurring and self.related:
+            original = self.related.pop()
+            self.rrule = original.rrule
```
Comment (WhyNotHugo) [inline todoman/model.py:266]: Isn't `original` the newer entry here?

File: tests/test_cli.py [hunk 5]
```diff
@@ -747,6 +748,55 @@ def test_done_recurring(runner, todo_factory, todos):
...
+    result = runner.invoke(cli, ["undo", "1"])
+    assert not result.exception
+
+    todo = next(todos(status="ANY"))
+    assert todo.percent_complete == 0
+    assert todo.is_completed is False
+    assert not todo.rrule
+
+
+@pytest.mark.xfail(reason="Not implemented")
```
Comment (WhyNotHugo) [inline tests/test_cli.py:765]: I think that recursion is now implemented, right?

Comment (r4ulill0) [reply tests/test_cli.py:765]: No, it is not.

There are multiple PRs regarding recursion. The most relevant here is #288 but there are some more listed in #485.

File: todoman/cli.py (hunk 2 above)
Comment (r4ulill0) [reply todoman/cli.py]: It doesn't. Restored

File: todoman/model.py (hunk 4 above)
Comment (r4ulill0) [reply todoman/model.py]: Yeah, bad naming. Solved.

File: todoman/model.py [hunk 6]
```diff
@@ -250,6 +250,27 @@ def complete(self) -> None:
...
+
+        Marks it as needs action, resets the percentage to 0 and deletes the
+        completed_at datetime.
+        If this todo belongs to a series, the one created at completion will
+        be deleted.
+
+        Returns the todo that should be deleted.
+        """
+        recurred = None
+        if self.is_recurring and self.related:
```
Comment (WhyNotHugo) [inline todoman/model.py:264]: I ran:

```sh
todo new -l test test
todo new 1
```

The completed todo looks like this:

```icalendar
BEGIN:VCALENDAR
VERSION:2.0
PRODID:io.barrera.todoman
BEGIN:VTODO
COMPLETED:20231118T090119Z
CREATED:20231118T090107Z
DTSTAMP:20231118T090107Z
DUE:20231119T090107Z
LAST-MODIFIED:20231118T090119Z
PERCENT-COMPLETE:100
SEQUENCE:2
STATUS:COMPLETED
SUMMARY:test
UID:1d54c495739f4546a92974ab574b5347@destiny.whynothugo.nl
END:VTODO
END:VCALENDAR
```

It doesn't have a recurrence rule any more (that is moved to the next one in the series), so this condition does not match.

If I run `todo undo 1`, both todos are kept around.

This is not ideal, but I don't think that it is terrible either. The RFC doesn't even cover how a recurring TODO is completed, so there's no correct behavior to any of this either way.

However, this `if` will never run, so this code is misleading. I think it should be removed (unless I'm missing something here).

**Pull request closed without merging**